regression-tests/pure2-function-type-id.cpp2

```diff
@@ -0,0 +1,108 @@
+main: () = {
+  postfix_operators();
+
+  // Variables with type of a mix of `*`/`const` to `() throws -> void`.
+  f0: * () throws = :()       = {};
+  f1: const * () throws       = f0;
+  f2: * const () throws       = f0;
+  f3: const * const () throws = f0;
+
+  // Uninitialized.
+  f4: * () throws;
+  f4 = f0;
+
+  f10: * * () throws             = f0&;
+  f11: const * * () throws       = f10;
+  f12: * const * () throws       = f10;
+  f13: const * const * () throws = f10;
+
+  i: i32                = 0;
+  i0: * i32             = i&;
+  i1: const * i32       = i0;
+  i2: * const i32       = i0;
+  i3: const * const i32 = i0;
+
+  // Assert consistent '*'/'const' with non-function type variables.
+  static_assert((std::is_const_v<decltype(f10)>) == std::is_const_v<decltype(i0)>);
+  static_assert((std::is_const_v<decltype(f11)>) == std::is_const_v<decltype(i1)>);
+  static_assert((std::is_const_v<decltype(f12)>) == std::is_const_v<decltype(i2)>);
+  static_assert((std::is_const_v<decltype(f13)>) == std::is_const_v<decltype(i3)>);
+  _ = f10;
+  _ = f11;
+  _ = f12;
+  _ = f13;
+  _ = i0;
+  _ = i1;
+  _ = i2;
+  _ = i3;
+
+  // Variables with various kinds of parameter.
+  f5: * (_: i32) throws = :(x: i32) = {};
+  f6: * (_: std::any) throws = :(x: std::any) = {};
+  f7: * (move _: i32) throws = :(move x: i32) = {};
+  f8: * (out _: i32) throws = :(copy x) = {};
+
+  // In alternative.
+  [[assert: inspect f0 -> bool {
+    is ()          = (std::terminate(), false);
+    is () throws   = (std::terminate(), false);
+    is * () throws = true;
+    is _           = false;
+  }]]
+  [[assert: inspect f0* -> bool {
+    is () throws = true;
+    is _         = false;
+  }]]
+
+  // As block variable.
+  (f: * () throws = f0) { }
+  (f: () throws   = f0*) { }
+
+  // As local function parameter.
+  _ = :(f: * () throws) = {};
+  _ = :(f: * () throws -> * () throws) = {};
+  _ = :(f: * () throws -> * () throws -> * () throws) = {};
+
+  // In local function return type.
+  _ = :() -> * () throws = nullptr;
```

Comment:
This still fails:
```Cpp2
  _ = :() -> forward() throws = f0$*;
```
Lowers to:
```C++
  (void) [_0 = f0]() -> auto () -> void& { return *cpp2::assert_not_null(_0);  };
```
Instead of:
```C++
  (void) [_0 = f0]() -> std::type_identity_t<auto (&)() -> void> { return *cpp2::assert_not_null(_0);  };
```